common/flatpak-dir.c

```diff
@@ -4174,37 +4403,40 @@ repo_pull (OstreeRepo                           *self,
 
       /* Pull options */
       g_variant_builder_init (&pull_builder, G_VARIANT_TYPE ("a{sv}"));
-      get_common_pull_options (&pull_builder, ref_to_fetch, dirs_to_pull, current_checksum,
-                               force_disable_deltas, flags, progress);
+
+      /* TODO: Find a way to retrieve different pull options per remote */
+      get_common_pull_options (&pull_builder, self, NULL, NULL, ref_to_fetch,
+                               dirs_to_pull, current_checksum, force_disable_deltas,
+                               flags, progress, cancellable);
       pull_options = g_variant_ref_sink (g_variant_builder_end (&pull_builder));
 
       context = flatpak_main_context_new_default ();
 
       if (results_to_fetch == NULL)
         {
-          ostree_repo_find_remotes_async (self, (const OstreeCollectionRef * const *) collection_refs_to_fetch,
+          ostree_repo_find_remotes_async (repo, (const OstreeCollectionRef * const *) collection_refs_to_fetch,
```

Comment:
Note: There are other calls to `ostree_repo_find_remotes_async()` in flatpak which you’ll also need to modify to send the right tokens.

Reply:
Regarding sending the right headers to the right servers, the issue is that OSTree’s P2P support is deliberately designed to allow requesting many refs from many servers in parallel, to try and reduce the number of network requests. I don’t think it’s possible to set a single header which would serve as a suitable token for all the requests, so I think the solution has to be for the calling code in flatpak to provide OSTree with all the tokens it could possibly need for a single `ostree_repo_find_remotes_async()` call, and information on when to use each of them.

Since each token is valid for a single ref, I suggest you could implement a `ref-http-headers` option (that’s just one suggested name; feel free to come up with a better one) in OSTree which is a map of ref-name to an `a(ss)` array of headers to send whenever requesting the commit object for that ref-name.

This would need to be proposed in upstream OSTree, need implementing there, and then it could be used in flatpak.